### PR TITLE
Update qod api documentation to 0.8.0

### PR DESCRIPTION
**Pull request by shilpa-padgaonkar**

### CODE REVIEW

File: documentation/API_documentation/QoD_API.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Overview
+
+<span class="colour" style="color:rgb(0, 0, 0)">The Quality-On-Demand (QoD) API provides programmable interface for developers and other users (capabilities consumers) to request stable latency or throughput managed by Telco networks without the necessity to have an in-depth knowledge of the 4G/5G system or the overall complexity of the Telecom Systems [[1]](#1).</span> 
+
+## 1\. Introduction
+
+Industrial (IoT), VR/Gaming, broadcasting, autonomous driving and many others scenarios demand network communication quality and are sensitive to any change in transmission conditions. Being able to request a stable latency (reduced jitter) or stable throughput from the network can improve user experience substantially.
```
Comment (tlohmar) [inline documentation/API_documentation/QoD_API.md:7]:
```suggestion
Industrial (IoT), VR/Gaming, Live Video Streaming, autonomous driving and many other scenarios demand network communication quality and are sensitive to any change in transmission conditions. Being able to request a stable latency (reduced jitter) or stable throughput from the network can improve user experience substantially.
```

**Comment by patrice-conil** [comment]
Fine by me

File: documentation/API_documentation/QoD_API.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
...oken (optional):** Authentication token for callback API. e.g. c8974e592c2fa383d4a3960714 <br><br>**Response**<br> **200: Session created**<br>  Response body:<br> **duration:** Session duration in seconds.<br> **ueId:** The identifier of the user equipment <br>   **asId:** The identifer of the application server.<br>   **uePorts (optional):** The requested port(s) on the user equipment.<br>   **asPorts (optional):** The requested port(s) on the application server.<br>  **qos:** Qualifier of the requested throughput profile.<br>   **notificationUri (optional):** URI of the callback receiver.<br>   **notificationAuthToken (optional):** Authentication token for callback API.<br>   **id:** Session ID in UUID format.<br> e.g. 123e4567-e89b-12d3-a456-426614174000<br>   **startedAt:** Timestamp of session start, in seconds since Unix epoch.<br> e.g. 1639479600<br>   **expiresAt**: Timestamp of session expiration if the session was not deleted, in seconds since Unix epoch. e.g. 1639566000 <br><br> **400:** **Invalid input.**<br> **401:** **Un-authorized. <br> **403:** Forbidden.**<br> **409:** **Conflict.**<br> **500:** **Server Error.**<br> **503:** **Service temporarily unavailable.** |
```
Comment (tlohmar) [inline documentation/API_documentation/QoD_API.md:99]:
```suggestion
| **HTTP Request**<br> POST \<base-url>/qod-api/v0/sessions<br>**Query Parameters**<br> No query parameters are defined.<br>**Path Parameters**<br> No path parameters are defined.<br>**Request Body Parameters**<br> **duration (optional)**: Session duration in seconds. Maximal value of 24 hours is used if not set. e.g. 86400<br> **ueId:** The identifier for the user equipment(device). The developer can choose to provide the below specified user equipment identifiers:<br>  - IPv4 address (supports mask) e.g. 192.168.0.1/24<br> - ipv6 address (supports mask) e.g. 2001:db8:85a3:8d3:1319:8a2e:370:7344<br> - msisdn (including country code and optionally could be prefixed by "+" sign) e.g. 004912345678923 <br> - externalId [[5]](#5)  assigned by the Mobile network Operator for the user equipment. e.g. 123456789@domain.com <br>NOTE: the communication service provider (CSP) might support only a subset of these options. The API invoker can provide multiple identifiers to be compatible across different CSPs. In this case the identifiers MUST belong to the same UE<br> **asId:** The identifier used for application server. The developer can choose from the below application server identifiers: <br> - ipv4 address (supports mask) e.g. 192.168.0.1/24 <br> - ipv6 address (supports mask) e.g. 2001:db8:85a3:8d3:1319:8a2e:370:7344 <br> **uePorts (optional):** A list of single ports or port ranges on the user equipment.<br> e.g. "uePorts": {"ranges": [{"from": 5010,"to": 5020}],"ports": [5060,5070]} <br>  **asPorts (optional):** A list of single ports or port ranges on the application server. e.g. "asPorts": {"ranges": [{"from": 5010,"to": 5020}],"ports": [5060,5070]}<br>  **qos:** Qualifier for the requested latency/throughput profile. e.g. QOS_E <br> **notificationUri (optional):** URI of the callback receiver. Allows asynchronous delivery of session related events. e.g. '[https://application-server.com/notifications](https://application-server.com/notifications)'</span><br> **notificationAuthToken (optional):** Authentication token for callback API. e.g. c8974e592c2fa383d4a3960714 <br><br>**Response**<br> **200: Session created**<br>  Response body:<br> **duration:** Session duration in seconds.<br> **ueId:** The identifier of the user equipment <br>   **asId:** The identifer of the application server.<br>   **uePorts (optional):** The requested port(s) on the user equipment.<br>   **asPorts (optional):** The requested port(s) on the application server.<br>  **qos:** Qualifier of the requested throughput profile.<br>   **notificationUri (optional):** URI of the callback receiver.<br>   **notificationAuthToken (optional):** Authentication token for callback API.<br>   **id:** Session ID in UUID format.<br> e.g. 123e4567-e89b-12d3-a456-426614174000<br>   **startedAt:** Timestamp of session start, in seconds since Unix epoch.<br> e.g. 1639479600<br>   **expiresAt**: Timestamp of session expiration if the session was not deleted, in seconds since Unix epoch. e.g. 1639566000 <br><br> **400:** **Bad Request.**<br> **401:** **Unauthorized. <br> **403:** Forbidden.**<br> **409:** **Conflict.**<br> **500:** **Server Error.**<br> **503:** **Service temporarily unavailable.** |
```
The Status Codes are well defined.

File: documentation/API_documentation/QoD_API.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+### 4.4 Errors
+
+Since CAMARA QoD API is based on REST design principles and blueprints, well defined HTTP status codes and families specified by community are followed [[4]](#4).
+
+Details of HTTP based error/exception codes for the QoD API are described in Section 4.3 of each API REST based method.
+Following table provides an overview of common error names, codes, and messages applicable to QoD API.
+
+| No | Error Name | Error Code | Error Message |
+| --- | ---------- | ---------- | ------------- |
+|1	|400 |	INVALID_INPUT |	"Expected property is missing: ueId.msisdn" |
```
Comment (tlohmar) [inline documentation/API_documentation/QoD_API.md:124]:
```suggestion
|1	|400 |	Bad Request|	"Expected property is missing: ueId.msisdn" |
```

Corrected Status Code name.

File: documentation/API_documentation/QoD_API.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+|2	|400 |	INVALID_INPUT |	"Expected property is missing: ueId.ipv4addr" |
+|3	|400 |	INVALID_INPUT |	"Expected property is missing: ueId.ipv4addr or ueId.ipv6addr" |
+|4	|400 |	INVALID_INPUT |	"Expected property is missing: uePorts" |
+|5	|400 |	INVALID_INPUT |	"Expected property is missing: qos" |
+|6	|400 |	INVALID_INPUT |	"Ranges not allowed: uePorts" |
+|7	|401 |	UNAUTHORIZED |	"No authorization to invoke operation" |
+|8	|403 |	FORBIDDEN |	"Operation not allowed" |
+|9	|404 |	NOT_FOUND |	"Session Id does not exist" |
+|10	|409 |	CONFLICT |	"Another session is created for the same UE" |
```
Comment (tlohmar) [inline documentation/API_documentation/QoD_API.md:133]: Question: I thought that an API developer can create multiple QoD sessions with different Flow Descriptions (e.g. different ports) to the same device.

File: documentation/API_documentation/QoD_API.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+|11	|500 |	INTERNAL |	"Session could not be created" |
```
Comment (tlohmar) [inline documentation/API_documentation/QoD_API.md:134]:
```suggestion
|11	|500 |	Internal Server Error |	"Session could not be created" |
```
1: The status code should be corrected. 
2: Does this error only occur, when the session cannot be created?

File: documentation/API_documentation/QoD_API.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+### 4.5 Policies
+
+N/A
+
+### 4.6 Code Snippets
+<br>
+<span class="colour" style="color:rgb(36, 41, 47)">Snippet 1, elaborates REST based API call with "*curl"* to create a QoS session for sample streaming service with following parameters: </span>
+
+* QoS session with QoS-profile "QOS_E" mapping,
+* App-Flow is specified for UE-Terminal identifier (ueId=2001:db8:85a3:8d3:1319:8a2e:370:7344), Application server identifier (asId=54.204.25.0/28) and Port number (asPorts=33001).
```
Comment (tlohmar) [inline documentation/API_documentation/QoD_API.md:146]: Looks complicated to have an IPv6 address as UE identifier and an IPv4 address as an Application Server Id. This may raise a bunch of technical questions.

File: documentation/API_documentation/QoD_API.md (hunk 3 above)
Comment (akoshunyadi) [reply documentation/API_documentation/QoD_API.md:124]: I guess this errorinfo.code should be a detailed error code, and not the text for the http response code.

File: documentation/API_documentation/QoD_API.md (hunk 1 above)
Comment (shilpa-padgaonkar) [reply documentation/API_documentation/QoD_API.md]: fixed

File: documentation/API_documentation/QoD_API.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+**Duration** 
+Number of seconds for which the QoD session should be created. This parameter is optional and if not specified, the session is either deleted on user request or if default expiration limit has been reached (24 hours in one of the example provider implementations).
+
+**Notification URL and token**
+Developers have a chance to specify callback URL on which notifications (eg. session termination) regarding the session can be received from the service provider. This is an optional parameter.
+
+Sample API invocations are presented in Section 4.6.
+
+## 3\. Authentication and Authorization
```
Comment (jlurien) [inline documentation/API_documentation/QoD_API.md:46]: We have to consider as well a three_legged securityScheme. I see it's not considered in the current 0.8 but it's used in our integrations. We can make a separate PR to add it explicitly.

**Review by jlurien — COMMENTED** [review]
For TEF it is important to add three_legged Authentication as a possibility. We will move a separate PR to add this

File: documentation/API_documentation/QoD_API.md (hunk 7 above)
Comment (tlohmar) [reply documentation/API_documentation/QoD_API.md:46]: @jlurien, all: Aha, I missed, that the IPv6 was introduced. Still, we should not mix usage of IPv4 and IPv6 in the same QoD Session. Either ueId and AsId are IPv4 or both are IPv6.

File: documentation/API_documentation/QoD_API.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+## 1\. Introduction
+
+Industrial (IoT), VR/Gaming, broadcasting, autonomous driving and many others scenarios demand network communication quality and are sensitive to any change in transmission conditions. Being able to request a stable latency (reduced jitter) or stable throughput from the network can improve user experience substantially.
+
+The QoD API offers the application developers the capability to request for stable latency (reduced jitter) or throughput for a specified App-Flow between User Equipment (application clients) and Application Servers (backend services). The developer has a pre-defined set of Quality of Service (QoS) Profiles which they could choose from depending on their latency or throughput requirements.
+
+<img src="./resources/QoD_overview.PNG" alt="QoD_LM" title="QoD API Overview">
```
[the remark this answers is not in the document]
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:11]: @tlohmar The network is actually not shown (only the backend system which offering the session recource)
@jlurien That's right but difficult to show without creating the impresson that both UE and AS have to communicate with the API.
Any proposal for a better illustration (not a formal correct picture) is appreciated.

File: documentation/API_documentation/QoD_API.md (hunk 6 above)
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:146]: I suggest that it is an actual, even typical case: Let's assume that the mobile network is internally using only IPv6 (or only the IPv6 address is unique to identify the UE), and the application server is offering the service on IPv4 address. The mobile network is then using CGNAT. I agree that there are a several technical questions which need to be addressed (partly candidates for the FAQ section, partly open issue, e.g. #34).

File: documentation/API_documentation/QoD_API.md (hunk 7 above)
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:46]: @tlohmar: See my comment above https://github.com/camaraproject/QualityOnDemand/pull/71#discussion_r1031883038

File: documentation/API_documentation/QoD_API.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (tlohmar) [reply documentation/API_documentation/QoD_API.md:11]: Ok, I am giving it a stab. I think, it would be good to show the QoS session (between the UE application and the Application Server, going through a 5G System) and then the interaction between API invoker (e.g. the application server) and the QoD function.

File: documentation/API_documentation/QoD_API.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+Security access keys such as OAuth 2.0 client credentials used by Client applications to invoke the QoD API.
+
+**QoS Profile**
+Latency or throughput requirements of the application mapped to relevant QoS profile class.
+
+**Identifier for the the user equipment (UE)**
+At least one identifier for the user equipment out of four options: IPv4 address, IPv6 address, MSISDN, or external ID [[5]](#5) assigned by the Mobile Operator for the user equipment
+
+**Identifier for the application server (AS)**
+IPv4 and/or IPv6 address of the application server (application backend)
```
[the remark this answers is not in the document]
Comment (tlohmar) [reply documentation/API_documentation/QoD_API.md:33]: Ok, i missed, that now IPv6 is also supported.

File: documentation/API_documentation/QoD_API.md (hunk 6 above)
Comment (tlohmar) [reply documentation/API_documentation/QoD_API.md:146]: You mean, the CGNAT is doing NAT64? I suggest to add some more description around the network assumptions. The Application Developers need to understand the concepts of the CAMARA QoD funtion.

**Comment by ToshiWakayama-KDDI** [comment]
Hi,
I raised some issues on Mapping table at the last QoD API meeting on 18th Nov, and it is supposed to be discussed by email discussion.  I would like Herbert to start email discussion and the Mapping table in the API document should be reflected.  
Thanks.

File: documentation/API_documentation/QoD_API.md (hunk 5 above)
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:134]: It's just reflecting the current YAML v0.8.0 ... so need to be changed there first and can then be adapted in the documentation.

File: documentation/API_documentation/QoD_API.md (hunk 4 above)
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:133]: @tlohmar Thanks for excellent observation ... need to be corrected with an issue and PR in the YAML first. What was actually meant (most probably) "There is an overlapping session existing for the same UE" ... but tbd.

File: documentation/API_documentation/QoD_API.md (hunk 6 above)
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:146]: @tlohmar: as said, this is a perfect example for an FAQ. The challenge is that there is no general explanation possible as operator network implementation vary. Do you see how this can be done in a general FAQ entry here in CAMARA?

**Comment by tlohmar** [comment]
I think, it is the Error Codes e.g. 400 is Bad Request (not Invalid input).

I don’t know, why it was not possible to comment on individual lines.

**Comment by hdamker** [comment]
>



> Hi,
> I raised some issues on Mapping table at the last QoD API meeting on 18th Nov, and it is supposed to be discussed by email discussion. I would like Herbert to start email discussion and the Mapping table in the API document should be reflected.
> Thanks.

@ToshiWakayama-KDDI Thanks for raising the issue with the mapping table which we will discuss as proposed. For the time being we deleted the copy of the table from the documentation and just kept the link.

File: documentation/API_documentation/QoD_API.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:11]: @tlohmar To get the updated documentation done for now, I suggest that you propose an updated picture in a minor editorial PR later.

File: documentation/API_documentation/QoD_API.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+## 4\. API Documentation
+
+### 4.1 API Version
+
+0.8.0
+
+### 4.2 Details
+
+The usage of the QoD API is based on QoS profile classes and parameters which define App-Flows.
+Based on the API, QoS sessions can be created, queried, and deleted. Once an offered QoS profile class is requested, application users get a prioritized service with stable latency or throughput even in the case of congestion. The QoD API has the following characteristics:
```
[the remark this answers is not in the document]
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md]: That should be fixed in most cases now

File: documentation/API_documentation/QoD_API.md (hunk 3 above)
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:118]: INVALID_INPUT is within the YAML v0.8.0 defined. If it should be changed than later together with the YAML.

File: documentation/API_documentation/QoD_API.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+* The prioritized App-Flow is described by providing information such as user equipment (UE) IP address (or other UE identifier) & application server (AS) IP addresses and port/port-ranges.
+* The developer can optionally specify the duration for which they need the prioritized App-flow.
+* Stable latency or throughput is requested by selecting from the list of QoS profiles made available by the service provider (e.g. QOS_E) to map latency and throughput requirements.
+* The developer can optionally also specify callback URL on which notifications for the session can be sent. <br>
+
+Following diagram shows the interaction between different components
+
+<br>
+
+<img src="./resources/QoD_details.PNG" alt="QoD_LM" title="QoD Management API">
```
[the remark this answers is not in the document]
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:73]: Fixed with https://github.com/camaraproject/QualityOnDemand/pull/71/commits/2d31e824f28a9b6c4392aa7a816231623189799e

**Comment by hdamker** [comment]
Thanks a lot to @tlohmar for your extensive and detailed review. We have incorporated as many suggestions as possible and especially fixed the points where something was wrong on the documentation side.

There are some comments open: 
- Some points which need first be proposed and discussed as changes to the API definition (e.g. error codes). The goal of this documentation update is to get consistent with version v0.8.0 of the API definition
- Other points need some more work and can be done independent of this update, e.g. a better illustration and adding FAQs

Said this, I propose that we try to merge this update as the base for further improvements
@tlohmar @jlurien @jpengar @eric-murray @patrice-conil @SyeddR @SylMOREL - would you approve?

**Review by hdamker — DISMISSED** [review]

**Comment by tlohmar** [comment]
Looks good. I dont have an "approve" button.

File: documentation/API_documentation/QoD_API.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
+# Overview
+
+<span class="colour" style="color:rgb(0, 0, 0)">The Quality-On-Demand (QoD) API provides programmable interface for developers and other users (capabilities consumers) to request stable latency or throughput managed by Telco networks without the necessity to have an in-depth knowledge of the 4G/5G system or the overall complexity of the Telecom Systems [[1]](#1).</span> 
+
+## 1\. Introduction
+
+Industrial (IoT), VR/Gaming, live video streaming, autonomous driving and many other scenarios demand network communication quality and are sensitive to any change in transmission conditions. Being able to request a stable latency (reduced jitter) or stable throughput from the network can improve user experience substantially.
```
Comment (jlurien) [inline documentation/API_documentation/QoD_API.md:7]:
```suggestion
Industrial (IoT), VR/Gaming, live video streaming, autonomous driving and many other scenarios demand network communication quality and are sensitive to any change in transmission conditions. Being able to request a stable latency (reduced jitter) or prioritized throughput from the network can improve user experience substantially.
```

**Review by jlurien — COMMENTED** [review]
Minor suggestion

File: documentation/API_documentation/QoD_API.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+Since CAMARA QoD API is based on REST design principles and blueprints, well defined HTTP status codes and families specified by community are followed [[4]](#4).
+
+Details of HTTP based error/exception codes for the QoD API are described in Section 4.3 of each API REST based method.
+Following table provides an overview of common error names, codes, and messages applicable to QoD API.
+
+| No | Error Name | Error Code | Error Message |
+| --- | ---------- | ---------- | ------------- |
+|1	|400 |	INVALID_INPUT |	"Expected property is missing: ueId.msisdn" |
+|2	|400 |	INVALID_INPUT |	"Expected property is missing: ueId.ipv4addr" |
+|3	|400 |	INVALID_INPUT |	"Expected property is missing: ueId.ipv4addr or ueId.ipv6addr" |
```
[the remark this answers is not in the document]
Comment (hdamker) [reply documentation/API_documentation/QoD_API.md:120]: @eric-murray I agree that the error definition are not the right ones for all implementations. But:
1. The commented errors are currently the ones as examples within the YAML file. We should change them in both locations together (together with other error infos). I will open an issue for that and collect the respective comments.
2. Some of the errors can make sense depending on the implementation. E.g. if the communication service provider expects either ipv4addr or ipv6addr then he error 3 would make sense (it doesn't matter if there are further ones provided by the API invoker)
My proposal is that operators adapt this part according to their implementation for now.

**Comment by hdamker** [comment]
@tlohmar: you should have an approve button now (added you to the reviewers)
@eric-murray: can you check if my answer is ok for you?
@jlurien: have commited your suggestion

**Comment by jpengar** [comment]
> @jlurien: have commited your suggestion

@hdamker thanks! ACK. Speaking on behalf of @jlurien (he is OOO this week).

**Review by eric-murray — APPROVED** [review]
